powerauth-push-server/src/main/java/io/getlime/push/service/PushSendingWorker.java

```diff
@@ -106,78 +112,112 @@ FcmClient prepareFcmClient(String serverKey) {
+                        callback.didFinishSendingMessage(PushSendingCallback.Result.PENDING);
+                        return;
+
+                    case FcmErrorResponse.MISMATCHED_CREDENTIAL:
+                    case FcmErrorResponse.INVALID_APNS_CREDENTIALS:
+                    case FcmErrorResponse.INVALID_ARGUMENT:
+                    case FcmErrorResponse.UNKNOWN_ERROR:
+                        Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.SEVERE, "Push message rejected by FCM gateway, error: " + errorCode);
+                        callback.didFinishSendingMessage(PushSendingCallback.Result.FAILED);
+                        return;
+
+                    default:
+                        Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.SEVERE, "Unexpected error code received from FCM gateway: " + errorCode);
+                        callback.didFinishSendingMessage(PushSendingCallback.Result.FAILED);
+                        return;
+                }
+            }
+
+            // Unexepected errors
+            Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.SEVERE, "Unexpected error occurred while sending push message: " + t.getMessage(), t);
+            callback.didFinishSendingMessage(PushSendingCallback.Result.FAILED);
         };
 
         // Perform request to FCM asynchronously, either of the consumers is called in case of success or error
-        fcmClient.exchange(request, onSuccess, onError);
+        try {
+            fcmClient.exchange(message, false, onSuccess, onError);
+        } catch (FcmMissingTokenException ex) {
+            Logger.getLogger(PushMessageSenderService.class.getName()).log(Level.SEVERE, ex.getMessage(), ex);
+            callback.didFinishSendingMessage(PushSendingCallback.Result.FAILED);
+        }
+    }
+
+    /**
+     * Build Android Message object from Push message body.
+     * @param pushMessageBody Push message body.
+     * @param attributes Push message attributes.
+     * @param pushToken Push token.
+     * @return Android Message object.
+     */
+    private Message buildAndroidMessage(final PushMessageBody pushMessageBody, final PushMessageAttributes attributes, final String pushToken) {
+        // convert data from Map<String, Object> to Map<String, String>
+        Map<String, Object> extras = pushMessageBody.getExtras();
+        Map<String, String> data = new LinkedHashMap<>();
+        if (extras != null) {
+            for (Map.Entry<String, Object> entry : extras.entrySet()) {
+                data.put(entry.getKey(), entry.getValue().toString());
+            }
+        }
+
+        AndroidConfig.Builder androidConfigBuilder = AndroidConfig.builder()
+                .setCollapseKey(pushMessageBody.getCollapseKey());
+
+        if (pushServiceConfiguration.isFcmDataNotificationOnly()) { // notification only through data map
+            String notificationString = pushMessageBody.getTitle()
```

Comment:
Let's use multiple keys: `_notification.title`, ... etc.

Reply:
And let's do an experiment by still using key `_notification` with the `Object` payload.